File: java/src/main/java/benchmarks/JsonParseDom.java
package benchmarks;

import com.alibaba.fastjson2.JSONArray;
import com.alibaba.fastjson2.JSONObject;

public class JsonParseDom extends Benchmark {
    private String text;
    private long resultVal;

    @Override
    public String name() {
        return "JsonParseDom";
    }

    @Override
    public void prepare() {
        JsonGenerate generator = new JsonGenerate();
        generator.n = (int) configVal("coords");
        generator.prepare();
        generator.run(0);
        text = generator.getText();
    }

    private double[] calc(String text) {
        JSONObject json = JSONObject.parseObject(text);
        JSONArray coordinates = json.getJSONArray("coordinates");

        double x = 0.0, y = 0.0, z = 0.0;

        for (int i = 0; i < coordinates.size(); i++) {
            JSONObject coord = coordinates.getJSONObject(i);
            x += coord.getDoubleValue("x");
            y += coord.getDoubleValue("y");
            z += coord.getDoubleValue("z");
        }

        double len = coordinates.size();
        return new double[] {x / len, y / len, z / len};
    }

    @Override
    public void run(int iterationId) {
        double[] values = calc(text);
        resultVal += ((int)Helper.checksumF64(values[0]) + (int)Helper.checksumF64(values[1]) + (int)Helper.checksumF64(values[2])) & 0xFFFFFFFFL;
    }

    @Override
    public long checksum() {
        return resultVal;
    }
}

File: scala/src/main/scala/benchmarks/JsonParseDom.scala
package benchmarks

import com.alibaba.fastjson2.{JSONArray, JSONObject}

class JsonParseDom extends Benchmark:
  private var text: String = _
  private var resultVal: Long = 0L

  override def name(): String = "JsonParseDom"

  override def prepare(): Unit =
    val generator = new JsonGenerate()
    generator.n = configVal("coords").toInt
    generator.prepare()
    generator.run(0)
    text = generator.getText
    resultVal = 0L

  private def calc(text: String): (Double, Double, Double) =
    val json = JSONObject.parseObject(text)
    val coordinates = json.getJSONArray("coordinates")

    var x = 0.0
    var y = 0.0
    var z = 0.0

    var i = 0
    while i < coordinates.size() do
      val coord = coordinates.getJSONObject(i)
      x += coord.getDouble("x")
      y += coord.getDouble("y")
      z += coord.getDouble("z")
      i += 1

    val len = coordinates.size()
    (x / len, y / len, z / len)

  override def run(iterationId: Int): Unit =
    val (x, y, z) = calc(text)

    val sum = (Helper.checksumF64(x) + Helper.checksumF64(y) + Helper.checksumF64(z)) & 0xffffffffL
    resultVal += sum

  override def checksum(): Long = resultVal & 0xffffffffL
